cmd/pulumi-test-language/providers/plain_provider.go

```diff
@@ -0,0 +1,354 @@
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+// A small provider with a single resource "Resource" that takes all its inputs via plain properties.
+// It includes nested types "Data" and "InnerData" that are also plain.
+type PlainProvider struct {
+	plugin.UnimplementedProvider
+}
+
+var _ plugin.Provider = (*PlainProvider)(nil)
+
+func (p *PlainProvider) Close() error {
+	return nil
+}
+
+func (p *PlainProvider) Configure(
+	context.Context, plugin.ConfigureRequest,
+) (plugin.ConfigureResponse, error) {
+	return plugin.ConfigureResponse{}, nil
+}
+
+func (p *PlainProvider) Pkg() tokens.Package {
+	return "plain"
+}
+
+func (p *PlainProvider) GetPluginInfo(context.Context) (workspace.PluginInfo, error) {
+	ver := semver.MustParse("13.0.0")
+	return workspace.PluginInfo{
+		Version: &ver,
+	}, nil
+}
+
+func (p *PlainProvider) GetSchema(
+	context.Context, plugin.GetSchemaRequest,
+) (plugin.GetSchemaResponse, error) {
+	makePrimitive := func(t string) schema.PropertySpec {
+		return schema.PropertySpec{
+			TypeSpec: schema.TypeSpec{
+				Type:  t,
+				Plain: true,
+			},
+		}
+	}
+	typeProperties := map[string]schema.PropertySpec{
+		"boolean": makePrimitive("boolean"),
+		"float":   makePrimitive("number"),
+		"integer": makePrimitive("integer"),
+		"string":  makePrimitive("string"),
+		"boolArray": {
+			TypeSpec: schema.TypeSpec{
+				Type: "array",
+				Items: &schema.TypeSpec{
+					Type:  "boolean",
+					Plain: true,
+				},
+				Plain: true,
+			},
+		},
+		"stringMap": {
+			TypeSpec: schema.TypeSpec{
+				Type: "object",
+				AdditionalProperties: &schema.TypeSpec{
+					Type:  "string",
+					Plain: true,
+				},
+				Plain: true,
+			},
+		},
+	}
+	typeRequired := []string{"boolean", "float", "integer", "string", "boolArray", "stringMap"}
+
+	dataProperties := maps.Clone(typeProperties)
+	dataProperties["innerData"] = schema.PropertySpec{
+		TypeSpec: schema.TypeSpec{
+			Type:  "ref",
+			Ref:   "#/types/plain:index:InnerData",
+			Plain: true,
+		},
+	}
+	dataRequired := append(typeRequired, "innerData")
+
+	resourceProperties := map[string]schema.PropertySpec{
+		"data": {
+			TypeSpec: schema.TypeSpec{
+				Type:  "ref",
+				Ref:   "#/types/plain:index:Data",
+				Plain: true,
+			},
+		},
+	}
+	resourceRequired := []string{"data"}
+
+	pkg := schema.PackageSpec{
+		Name:    "plain",
+		Version: "13.0.0",
+		Types: map[string]schema.ComplexTypeSpec{
+			"plain:index:InnerData": {
+				ObjectTypeSpec: schema.ObjectTypeSpec{
+					Type:       "object",
+					Properties: typeProperties,
+					Required:   typeRequired,
+				},
+			},
+			"plain:index:Data": {
+				ObjectTypeSpec: schema.ObjectTypeSpec{
+					Type:       "object",
+					Properties: dataProperties,
+					Required:   dataRequired,
+				},
+			},
+		},
+		Resources: map[string]schema.ResourceSpec{
+			"plain:index:Resource": {
```

Comment:
It's a little unusual to have plain values for a custom resource (they're primarily for components), but I guess someone may want to do it and it should work, so might as well test it. 👍

Reply:
Good point, we should check this works as a component as well.